ensime_shared/ensime.py

```diff
@@ -109,13 +111,13 @@ def create_client(self, config_path):
 
         This will launch the ENSIME server for the project as a side effect.
         """
-        server_v2 = self.using_server_v2()
-        editor = Editor(self._vim)
-        launcher = EnsimeLauncher(self._vim, config_path, server_v2)
-        if server_v2:
-            return EnsimeClientV2(editor, self._vim, launcher)
+        config = ProjectConfig(config_path)
+        launcher = EnsimeLauncher(self._vim, config)
+
+        if self.using_server_v2:
```

Comment:
You're right, it isn't ideal to keep a client config setting for this if it's now configurable from the build tools, I hadn't really thought about removing the client setting altogether.

I'm divided about how to go about this, though. Right now the launcher doesn't actually care what version it's launching, and it's kind of nice that it doesn't need to. Parsing it out of jar file names in `:ensime-server-jars` feels icky to me; it's not otherwise directly available as a field in the `.ensime` file, and introducing coupling to particular build tools to get it (e.g. asking sbt for `ensimeVersion`) seems wrong too, at least for this purpose.

IMO what feels right is to get the server version strictly from the handshake messages (`ConnectionInfo` or whatever might possibly come out of ensime/ensime-server#1552 in the future, it appears that hasn't been resolved yet). That's cleanly agnostic to the launcher strategy.

This implies that handshake is completed before an `EnsimeClient` gets instantiated for the appropriate protocol version. That would be a change (subject of a new enhancement, if we agree on this design), but it sounds like a sensible one to me. It would get trickier if this becomes async, but how much so is hard to say since those specifics aren't decided/implemented.

Maybe I'm making too big a deal out of avoiding stuff like parsing version out of the jars… clearly it isn't *hard* to do that, I'm just afraid it leads to multiple launch strategy solutions that are somewhat fragile, and internal coupling that can be avoided—I think there might be conscious reasoning in the fact that server version isn't expressly encoded in `.ensime`, like that a running server should be the authoritative source of truth for protocol version, not `.ensime`, nor the build tool, nor a launcher that's finished its job of opaquely invoking `java` with a classpath.

Thoughts?

If this sounds like the way to go, then I'll create a new ticket for making client protocol selection automatic based on handshake, and `g:ensime_server_v2` will be removed once that gets done.